clusterloader2/pkg/measurement/common/loadbalancer_nodesync_latency.go

```diff
@@ -0,0 +1,274 @@
+// If namespace is not passed by parameter, all LoadBalancer type service with all-namespace scope is assumed.
+// "measure" action triggers nodesync and observation of nodesync completion for selected LB services.
+// "gather" returns node sync latency summary.
+func (s *LoadBalancerNodeSyncMeasurement) Execute(config *measurement.Config) ([]measurement.Summary, error) {
+	s.client = config.ClusterFramework.GetClientSets().GetClient()
+	action, err := util.GetString(config.Params, "action")
+	if err != nil {
+		return nil, err
+	}
+	switch action {
+	case "measure":
+		if err := s.selector.Parse(config.Params); err != nil {
+			return nil, err
+		}
+		s.waitTimeout, err = util.GetDurationOrDefault(config.Params, "waitTimeout", defaultNodeSyncLatencyTimeout)
+		if err != nil {
+			return nil, err
+		}
+		return nil, s.measureNodeSyncLatency()
+	case "gather":
+		if err := s.labelNodeForLBs(false); err != nil {
+			return nil, err
+		}
+		return s.gather(config.Identifier)
+	default:
+		return nil, fmt.Errorf("unknown action %v", action)
+	}
+}
+
+func (s *LoadBalancerNodeSyncMeasurement) Dispose() {}
+
+func (s *LoadBalancerNodeSyncMeasurement) String() string {
+	return loadBalancerNodeSyncLatencyName + ": " + s.selector.String()
+}
+
+func (s *LoadBalancerNodeSyncMeasurement) measureNodeSyncLatency() error {
+	ctx := context.Background()
+	options := metav1.ListOptions{}
+	s.selector.ApplySelectors(&options)
+	svcList, err := s.client.CoreV1().Services(s.selector.Namespace).List(ctx, options)
+	if err != nil {
+		return err
+	}
+
+	// lbSvcMap contains key->service mapping for load balancer type services
+	lbSvcMap := map[string]v1.Service{}
+	for _, svc := range svcList.Items {
+		if svc.Spec.Type == v1.ServiceTypeLoadBalancer {
+			key, err := runtimeobjects.CreateMetaNamespaceKey(&svc)
+			if err != nil {
+				return fmt.Errorf("meta key created error: %v", err)
+			}
+
+			lbSvcMap[key] = svc
+		}
+	}
+	totalLbSvc := len(lbSvcMap)
+
+	// trigger node sync by picking a node and add exclude lb label
+	nodeList, err := s.client.CoreV1().Nodes().List(ctx, metav1.ListOptions{})
+	if err != nil {
+		return err
+	}
+
+	for _, node := range nodeList.Items {
+		if !isCandidateNode(node) {
+			continue
+		}
+		s.excludedNodeName = node.Name
+		break
+	}
+
+	if s.excludedNodeName == "" {
+		return fmt.Errorf("failed to find a node candidate to trigger nodesync from node list: %v", nodeList.Items)
+	}
+	if err = s.labelNodeForLBs(true); err != nil {
+		return err
+	}
+
+	now := time.Now()
+	for key := range lbSvcMap {
+		s.svcNodeSyncLatencyTracker.Set(key, phaseNodeSyncStart, now)
+	}
+
+	return wait.Poll(5*time.Second, s.waitTimeout, func() (bool, error) {
+		for key, svc := range lbSvcMap {
+			if _, found := s.svcNodeSyncLatencyTracker.Get(key, phaseNodeSyncComplete); found {
+				continue
+			}
+			selector := s.client.CoreV1().Events(s.selector.Namespace).GetFieldSelector(&svc.Name, &svc.Namespace, nil, nil)
```

Comment:
Continuing the discussion about events.

> Consuming GCE API calls quota.

How many services you would eventually like to test? I would like to understand if simple bump of quota won't help here. 
Instead of getting target pools one by one - can we list (potentially in batches)?


The current event-based approach is also inefficient, because given that your `s.selector.Namespace` is effectively empty, we will be listing all events from the whole clusters for every service independently (i.e. the returned result will be small, but we will need to internally process and filter out all events for every call).

Reply:
I want to eventually test up to 500 LBs. (probably manual one time) And setup CI for 100 LBs. I have seen customers have >100 LBs. 

Listing TargetPools could not tell what instances are included. It requires a Get call to investigate the content and compare the instances one by one. 

I think clusterloader automatically creates separate namespaces for the load test. So `s.selector.Namespace` is only pointing to the test namespace. As I am using the feildSelector, the event Get call should only return events associated with the service, no?

Moreover, the event-based approach does not need to efficient, the retry interval can be configured to very high. This is because the latency measurement is depending on the `creationTimestamp` instead of the time when the validation took place.

Reply:
> I want to eventually test up to 500 LBs. (probably manual one time) And setup CI for 100 LBs. I have seen customers have >100 LBs.

OK - that doesn't sound that huge. On what size of cluster would you like to to test it?


> Listing TargetPools could not tell what instances are included. It requires a Get call to investigate the content and compare the instances against nodes one by one.

Got it.
I'm still a bit reluctant to using events - this is too whitebox-y to me. It's not future-proof, the fact that particular event was generated doesn't meen that LB works - it depends on the current implementation how the event is generated...

How sure are we that GCE API quota will be a problem? Can we mitigate it by choosing N random services and checking it for them? What if we would bump quotas?
[How many API calls do we need for updating nodes in a TP?]


> I think clusterloader automatically creates separate namespaces for the load test. 

Yes - but the selector.namespace is set to constant when you're parsing the config:
https://github.com/kubernetes/perf-tests/pull/1785/files#diff-dd38242128260b4a9cad8419f28cfb7f518618cd2e49c809fa532cfb4df36f4bR96
And you don't specify namespaces there - so it will be empty.
I'm not saying it wont work - I'm saying that it will be underneath listing all events in the system and filtering appropriate one.

> Moreover, the event-based approach does not need to efficient, the retry interval can be configured to very high. This is because the latency measurement is depending on the creationTimestamp instead of the time when the validation took place.

If we really need events (which I really hope we don't), a single informer sounds like a better option anyway:
https://github.com/kubernetes/perf-tests/blob/master/clusterloader2/pkg/measurement/util/informer/informer.go

Reply:
> I'm still a bit reluctant to using events - this is too whitebox-y to me. It's not future-proof, the fact that particular event was generated doesn't mean that LB works - it depends on the current implementation how the event is generated...

But checking TargetPool is also very whitebox-y. It only works for NetLB. For the classic L4 ILB, the event based approach also works. 

> How sure are we that GCE API quota will be a problem? 

The GCE API read quota is per project and across all APIs. By default, the quota is 20 QPS. I think it can be bumped but I only have impression with it bumped to 40 QPS at most. For 500 TPs, it took at least (assuming 20 QPS), 25s to traverse all of them. 

> Can we mitigate it by choosing N random services and checking it for them? What if we would bump quotas? [How many API calls do we need for updating nodes in a TP?]

This particular measurement is measuring how long would it take for service controller to update all LBs in the cluster when a nodesync got triggered. So it has to scan thru all LBs to be accurate. 

For each TP update, it actually consists of multiple GCE API calls. 1 List Instance , 1 TargetPool Add instances, 1 TargetPool delete instances, 2 Get TargetPool. So it is quite a few calls already. 


> If we really need events (which I really hope we don't), a single informer sounds like a better option anyway:
https://github.com/kubernetes/perf-tests/blob/master/clusterloader2/pkg/measurement/util/informer/informer.go

I can change it to use informer if efficiency is the concern here.

Reply:
> But checking TargetPool is also very whitebox-y. It only works for NetLB. For the classic L4 ILB, the event based approach also works.

Can we use something different for ILB?
Maybe pinging the service and checking when it starts to reach the new node (we would have to add node instead of deleting it instead)?
I would really like to avoid event dependency...

>  For 500 TPs, it took at least (assuming 20 QPS), 25s to traverse all of them.

FYI - We're bumping it to 200qps in our largest projects.

Reply:
For ILB(Subsetting), I am planning to add a new measurement similar to [exec](https://github.com/kubernetes/perf-tests/blob/05b63b33bd15916ca92f9c724cc8ed83eca6a056/clusterloader2/pkg/measurement/common/exec.go#L47) so we can list the NEGs and verify the node names to be what we expect.

The presence of event can be used to trigger the NEG measurement, so we do not check the contents until necessary. We could omit it too. 

For IG-based ILB, we would need to list contents of the single instance group, since it is shared.
Perhaps we need very similar measurements for each case - NEG, IG, TargetPool. For IG and TargetPool, the steps are - 1) List all cluster nodes 2) list all instances in GCE resource. Ensure they have the same entries.

NEG is trickier since it uses a subset.

Reply:
Let me summarize and compare the 2 approaches:

1. Event based approach (this PR)
   - Simple to implement
   - Latency measurement is accurate (from Nodesync starts and service controller finishes update LB hosts and emits event)
   - Can be used for both classic ILB/NetLB. 
   - Implementation specific (depending on the OSS service controller emitting the event)

2. TargetPool based approach
  - Polling based
  - Measurement is dependent on the frequency of the observation. (needs high RPS to )
  - Implementation specific (Only fits classic NetLB with Target Pool)


There are essentially 2 ways to implement the TargetPool based measurement:

A. Use Exec measurement as building block to run gcloud commands, the rough steps will be: 
  - Use the NetLB naming scheme and generate TargetPool name
  - Find the region
  - Run `gcloud to run gcloud compute target-pool describe ${targetpool} --region ${region}` (There is no list-istance gcloud API on targetPool)
  - Then do string parsing on the return output to match up with the expected nodes.

B. Introduce GCE API go clients into clusterloader2. 
- Construct the parameters (Targetpool Name and region) and then execute 
- use GCE API go client to poll target pool instances and parse it via golang. 

A few more factors to consider:
1. The immediate goal of this load test is to measure the scalability improvements caused by the parallelized backend management. 
2. RegionalBackendService migration is going to kick start soon. New NetLBs will be based on RBS instead of target pools

Overall, I would suggest the event based approach is a good compromise before RBS is in place. And we can improve this measurement incrementally. WDYT?

Reply:
I'm definitely not  fan of relying on events, but I don't want to block this infinitely.

So for the sake of making progress, let's proceed with the event-based approach for now (to unblock your tests).
I would like to get rid of this eventually, but we may discuss that separately.

So let's just fix it to use informer instead of periodically listing and proceed with this.